hydra-node/src/Hydra/Ledger/Cardano.hs

```diff
@@ -95,712 +82,510 @@ cardanoLedger =
+
+-- FIXME: The network discriminant should really be a parameter here.
+mkVkAddress ::
+  VerificationKey PaymentKey ->
+  Address ShelleyAddr
+mkVkAddress vk =
+  makeShelleyAddress
+    (Testnet $ NetworkMagic 42)
+    (PaymentCredentialByKey $ verificationKeyHash vk)
+    NoStakeAddress
 
-  dpState = Cardano.DPState def def
+--
+-- Type conversions & plumbing
+--
 
-  -- NOTE(AB): This sets some parameters for the tx generator that will
-  -- affect the structure of generated trasactions. In our case, we want
-  -- to remove "special" capabilities which are irrelevant in the context
-  -- of a Hydra head
-  -- see https://github.com/input-output-hk/cardano-ledger-specs/blob/nil/shelley/chain-and-ledger/shelley-spec-ledger-test/src/Test/Shelley/Spec/Ledger/Generator/Constants.hs#L10
-  generatorEnv = (genEnv Proxy){geConstants = noPPUpdatesTransactions}
+--
+-- Utxo
+--
 
-genSequenceOfValidTransactions :: Utxo CardanoTx -> Gen [CardanoTx]
-genSequenceOfValidTransactions initialUtxo
-  | initialUtxo == mempty = pure []
-  | otherwise = do
-    n <- getSize
-    numTxs <- choose (1, n)
-    genFixedSizeSequenceOfValidTransactions numTxs initialUtxo
+type Utxo = Utxo' LedgerEra (Ledger.Alonzo.TxOut LedgerEra)
 
-genFixedSizeSequenceOfValidTransactions :: Int -> Utxo CardanoTx -> Gen [CardanoTx]
-genFixedSizeSequenceOfValidTransactions numTxs initialUtxo
-  | initialUtxo == mempty = pure []
-  | otherwise = do
-    reverse . snd <$> foldM newTx (initialUtxo, []) [1 .. numTxs]
- where
-  newTx (utxos, acc) _ = do
-    tx <- genCardanoTx utxos
-    case applyTransactions cardanoLedger utxos [tx] of
-      Left err -> error $ show err
-      Right newUtxos -> pure (newUtxos, tx : acc)
+-- | Newtype with phantom types mostly required to work around the poor interface
+-- of 'Ledger.UTXO'and provide 'Monoid' and 'Foldable' instances to make utxo
+-- manipulation bareable.
+newtype Utxo' era out = Utxo
+  { unUtxo :: Map (Ledger.TxIn (Ledger.Crypto era)) out
+  }
+  deriving newtype (Eq, Show)
 
-type CardanoEra = MaryEra StandardCrypto
+instance ToCBOR Utxo where
+  toCBOR = toCBOR . coerce @_ @(Ledger.UTxO LedgerEra)
+  encodedSizeExpr sz _ = encodedSizeExpr sz (Proxy @(Ledger.UTxO LedgerEra))
 
-type CardanoTxBody = Cardano.TxBody CardanoEra
+instance FromCBOR Utxo where
+  fromCBOR = coerce @(Ledger.UTxO LedgerEra) <$> fromCBOR
+  label _ = label (Proxy @(Ledger.UTxO LedgerEra))
 
-type CardanoTxWitnesses = Cardano.WitnessSet CardanoEra
+instance Functor (Utxo' era) where
+  fmap fn (Utxo u) = Utxo (fmap fn u)
 
-type CardanoAuxiliaryData = StrictMaybe (AuxiliaryData CardanoEra)
+instance Foldable (Utxo' era) where
+  foldMap fn = foldMap fn . unUtxo
+  foldr fn zero = foldr fn zero . unUtxo
 
-type CardanoAddress = Cardano.Addr StandardCrypto
+instance Semigroup Utxo where
+  Utxo uL <> Utxo uR = Utxo (uL <> uR)
 
-type CardanoKeyPair = Cardano.KeyPair 'Cardano.Payment StandardCrypto
+instance Monoid Utxo where
+  mempty = Utxo mempty
```

Comment:
Could: Not use explicit but `deriving newtype` instances for `Semigroup`and `Monoid`

Reply:
Hmmm. It wasn't possible before because I was wrapping `Ledger.UTxO`, but I later changed that so that's a good simplification indeed!